util/resolver/authorizer.go

```diff
@@ -166,6 +166,11 @@ func (a *dockerAuthorizer) AddResponses(ctx context.Context, responses []*http.R
 				}
 			}
 
+			// prevent cached expired (or otherwise newly invalid) tokens from being reused
+			if handler != nil && last.StatusCode == http.StatusUnauthorized {
```

Comment:
Couldn't we just add this condition to `invalidAuthorization` function or is there a difference?

Reply:
We could move the status check to `invalidAuthorization`, at first glance the check there didn't seem to be somewhere we would also check the last responses http status code.

Additionally, if one token causes a 401, any other tokens in the handler are also deleted. Which isn't the end of the world; but may cause extra authorization traffic on these expiration boundaries.

Reply:
Just checked out adding it in there. If we do it there, would need to do it before the error string check coming from the challenge info.

Reply:
> Additionally, if one token causes a 401, any other tokens in the handler are also deleted. Which isn't the end of the world; but may cause extra authorization traffic on these expiration boundaries.

As the handler is already per call site I think it is fine assumption to make. Usually, they are equal and I think in the error case it is more likely that if there were other scopes they are invalid now as well.